packages/graph/hash_graph/lib/graph/src/types/schema/helper.rs

```diff
@@ -0,0 +1,291 @@
+use serde::{Deserialize, Serialize};
+
+use crate::types::schema::{Validate, ValidationError};
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(
+    try_from = "ArrayRepr<T>",
+    tag = "type",
+    rename = "array",
+    rename_all = "camelCase"
+)]
+pub struct Array<T> {
+    items: T,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    min_items: Option<usize>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    max_items: Option<usize>,
+}
+
+impl<T> Array<T> {
+    /// Creates a new `Array` without validating.
+    #[must_use]
+    pub fn new_unchecked<A: Into<Option<usize>>, B: Into<Option<usize>>>(
+        items: T,
+        min_items: A,
+        max_items: B,
+    ) -> Self {
+        Self {
+            items,
+            min_items: min_items.into(),
+            max_items: max_items.into(),
+        }
+    }
+
+    /// Creates a new `Array` from the given `items`, `min_items` and `max_items`.
+    ///
+    /// # Errors
+    ///
+    /// - [`ValidationError`] if validation is failing.
+    pub fn new<A: Into<Option<usize>>, B: Into<Option<usize>>>(
+        items: T,
+        min_items: A,
+        max_items: B,
+    ) -> Result<Self, ValidationError> {
+        let array = Self::new_unchecked(items, min_items, max_items);
+        array.validate()?;
+        Ok(array)
+    }
+
+    #[must_use]
+    pub const fn items(&self) -> &T {
+        &self.items
+    }
+
+    #[must_use]
+    pub const fn min_items(&self) -> Option<usize> {
+        self.min_items
+    }
+
+    #[must_use]
+    pub const fn max_items(&self) -> Option<usize> {
+        self.max_items
+    }
+}
+
+impl<T> Validate for Array<T> {
+    fn validate(&self) -> Result<(), ValidationError> {
+        Ok(())
+    }
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+enum ArrayTypeTag {
+    Array,
+}
+
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+struct ArrayRepr<T> {
+    #[serde(rename = "type")]
+    _type: ArrayTypeTag,
+    items: T,
+    min_items: Option<usize>,
+    max_items: Option<usize>,
+}
+
+impl<T> TryFrom<ArrayRepr<T>> for Array<T> {
+    type Error = ValidationError;
+
+    fn try_from(array: ArrayRepr<T>) -> Result<Self, Self::Error> {
+        Self::new(array.items, array.min_items, array.max_items)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum OneOrMany<T> {
+    One(T),
+    Array(Array<T>),
```

Comment:
nit:
Should we call the enum variant `Many` instead?

Reply:
Calling the struct `OneOrArray` would be fine for me

Reply:
Also good with that. Could also think about `SingleOrArray` or `ValueOrArray` or similar

Reply:
Went with f21f3a4d95df017baab8414ae3c6d94fd8437d51